File: 0092. Reverse Linked List II/pointer.recursion.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        dummy = ListNode(-1, head);
        prev = dummy
        curr = head;
        c = 1;
        while c < left:
            prev = curr;
            curr = curr.next;
            c+=1;
        
        def reverse(r, idx):
            if idx == right:
                return (r, r.next);
            
            newHead = reverse(r.next, idx+1)
            r.next.next = r;
            r.next = None;
            
            return newHead;a
        
        (newHead, remain) = reverse(curr, c)
        
        prev.next = newHead;
        curr.next = remain;
        
        return dummy.next;
                

File: 0092. Reverse Linked List II/pointer.recursion.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} left
 * @param {number} right
 * @return {ListNode}
 */
 var reverseBetween = function(head, left, right) {
    
    let remain = null;
    
    let c = 1;
    let dummy = new ListNode(-1, head)
    let curr = head, prev = dummy;
    while(c < left){
        prev = curr
        curr = curr.next;
        c++
    }
    
    const reverse = (r, idx) => {
        if(idx === right){
            remain = r.next;
            return r;
        }
        
        const newHead = reverse(r.next, idx+1);
        r.next.next = r;
        r.next = null;
        return newHead;
    }
    
    const newHead = reverse(curr, c);
    
    prev.next = newHead;
    curr.next = remain;
    
    return dummy.next;
    
};